imap/reconstruct.c

```diff
@@ -715,3 +731,94 @@ static void do_mboxlist(void)
     fprintf(stderr, "reconstructing mailboxes.db currently not supported\n");
     exit(EX_USAGE);
 }
+
+/*
+ * Reconstruct an mbentry from a mailbox header path.
+ */
+static void reconstruct_mbentry(const char *header_path)
+{
+    char real[PATH_MAX+1];
+    mbentry_t *mbentry = NULL, *mbentry_byname = NULL, *mbentry_byid = NULL;
+    struct mboxlock *namespacelock = NULL;
+
+    /* Create an mbentry from cyrus.header */
+    mbentry = mailbox_mbentry_from_path(realpath(header_path, real));
+    if (!mbentry) {
+        printf("Invalid/missing cyrus.header at %s\n", real);
+        return;
+    }
+    else if (!mbentry->partition) {
+        printf("header for %s is not located on a valid [meta]partition\n",
+               mbentry->name);
+        goto done;
+    }
+
+    namespacelock = mboxname_usernamespacelock(mbentry->name);
+
+    /* Look for existing N record and sanity check partition and uniqueid */
+    mboxlist_lookup_allow_all(mbentry->name, &mbentry_byname, NULL);
+    if (mbentry_byname) {
+        if (strcmpsafe(mbentry->partition, mbentry_byname->partition)) {
+            printf("partition clash with %s - %s vs %s\n",
+                   mbentry->name, mbentry->partition, mbentry_byname->partition);
+            goto done;
+        }
+
+        if (strcmpsafe(mbentry->uniqueid, mbentry_byname->uniqueid)) {
+            printf("uniqueid clash with %s - %s vs %s\n",
+                   mbentry->name,
+                   mbentry->uniqueid, mbentry_byname->uniqueid);
+            goto done;
+        }
```

Comment:
If either (but not both) of the header file or the N record are missing the uniqueid field, this is going to just bail out instead of fixing it.

If both places are missing uniqueid, it won't bail out (`"" == ""`), but I suspect it might crash later somewhere.

Reply:
Resolved by using cyrus.header, N record, makeuuid(), in that order